src/logic/map_objects/immovable_program.cc

```diff
@@ -110,94 +201,152 @@ void ImmovableProgram::ActPlaySound::execute(Game& game, Immovable& immovable) c
+		}
 	} catch (const WException& e) {
 		throw GameDataError("transform: %s", e.what());
 	}
 }
 
 void ImmovableProgram::ActTransform::execute(Game& game, Immovable& immovable) const {
-	if (immovable.apply_growth_delay(game)) {
-		return;
-	}
-	if (probability == 0 || game.logic_rand() % 256 < probability) {
-		Player* player = immovable.get_owner();
-		Coords const c = immovable.get_position();
-		MapObjectDescr::OwnerType owner_type = immovable.descr().owner_type();
-		immovable.remove(game);  //  Now immovable is a dangling reference!
+	assert(removal_wanted_ || !type_name_.empty());
 
-		if (bob) {
-			game.create_ship(c, type_name, player);
-		} else {
-			game.create_immovable_with_name(
-			   c, type_name, owner_type, player, nullptr /* former_building_descr */);
-		}
-	} else {
-		immovable.program_step(game);
-	}
-}
-
-ImmovableProgram::ActGrow::ActGrow(std::vector<std::string>& arguments,
-                                   const ImmovableDescr& descr) {
-	if (arguments.size() != 1) {
-		throw GameDataError("Usage: grow=<immovable name>");
+	// Check whether we want to remove
+	if (removal_wanted_ &&
+	    (removal_probability_ == 0 || game.logic_rand() % 256 < removal_probability_)) {
+		immovable.remove(game);  //  Now immovable is a dangling reference!
+		return;
 	}
-	if (!descr.has_terrain_affinity()) {
-		throw GameDataError(
-		   "Immovable %s can 'grow', but has no terrain_affinity entry.", descr.name().c_str());
+	// For removal with chance but no transformation to new object
+	if (type_name_.empty()) {
+		return;
 	}
 
-	// TODO(GunChleoc): If would be nice to check if target exists, but we can't guarantee the load
-	// order. Maybe in postload() one day.
-	type_name = arguments.front();
-}
-
-void ImmovableProgram::ActGrow::execute(Game& game, Immovable& immovable) const {
 	if (immovable.apply_growth_delay(game)) {
 		return;
 	}
 
+	// Transform according to success if this was specified, and using any available terrain affinity
+	// if not.
 	const Map& map = game.map();
-	FCoords const f = map.get_fcoords(immovable.get_position());
 	const ImmovableDescr& descr = immovable.descr();
+	Player* player = immovable.get_owner();
+	FCoords const f = map.get_fcoords(immovable.get_position());
+	MapObjectDescr::OwnerType owner_type = immovable.descr().owner_type();
 
-	if ((game.logic_rand() % TerrainAffinity::kPrecisionFactor) <
-	    probability_to_grow(descr.terrain_affinity(), f, map, game.world().terrains())) {
-		MapObjectDescr::OwnerType owner_type = descr.owner_type();
-		Player* owner = immovable.get_owner();
+	const bool will_transform =
+	   descr.has_terrain_affinity() && transform_probability_ == 0 ?
+	      (game.logic_rand() % TerrainAffinity::kPrecisionFactor) <
+	         probability_to_grow(descr.terrain_affinity(), f, map, game.world().terrains()) :
+	      transform_probability_ == 0 || game.logic_rand() % 256 < transform_probability_;
```

Comment:
> There is a proposal in the issue to determine remove from the terrain affinity. I have not addressed that here, since probability calculus is not my strong suit.

Then what about something like:
```
bool will_transform = true;
if (descr.has_terrain_affinity()) {
	if ((game.logic_rand() % TerrainAffinity::kPrecisionFactor) >= probability_to_grow(descr.terrain_affinity(), f, map, game.world().terrains()) {
		if (transform_probability_) { will_transform = false; }
		else { immovable.remove(game); return; }
	}
} else {
	will_transform = transform_probability_ == 0 || game.logic_rand() % 256 < transform_probability_;
}
```

Reply:
What I mean is that currently, the removal probability is 100% hard-coded and always the same for a tree species, independent of terrain type. If we change this, it will need rebalancing. So, I'd rather not touch it in this branch.